src/expr/src/relation/mod.rs

```diff
@@ -2757,19 +2757,30 @@ impl RowSetFinishing {
         // let return_size = std::cmp::min(total, limit);
         //
         // but it breaks early if the limit is reached, instead of scanning the entire code.
-        let return_size = rows[offset_nth_row..]
+        let (num_rows, num_bytes) = rows[offset_nth_row..]
             .iter()
-            .try_fold(0, |sum, (_, count)| {
-                let new_sum = sum + count.get();
-                if new_sum > limit {
+            .try_fold((0, 0), |(num_rows, num_bytes), (row, count)| {
+                let new_rows = num_rows + count.get();
+                let new_bytes = num_bytes + row.byte_len();
+
+                // if we have more rows or more bytes than we can handle, bail
+                if new_rows > limit || new_bytes > max_result_size {
                     None
                 } else {
-                    Some(new_sum)
+                    Some((new_rows, new_bytes))
                 }
             })
-            .unwrap_or(limit);
+            .unwrap_or((limit, usize::MAX));
```

Comment:
I'm not sure that this is correct. If `new_rows > limit`, then `num_bytes` may still be less than `max_result_size`. In that case it would be something like `limit * size_of_rows`, however in this code we will have returned `usize::MAX`.

Reply:
Great catch!

Talked offline, updated to not use the `try_fold(...)` and instead `return` and `break` which should give us everything we want